### ISSUE TITLE
k means implementation

### ISSUE THREAD

**Issue by rola93** [OlafenwaMoses/ImageAI#337]
I've been reading the code, and I found that, at some point, I don't know why, but there is a kmeans function. It is defined [here](https://github.com/OlafenwaMoses/ImageAI/blob/7a8a6012a066dc50b5db64f12885aaddd96f4e97/imageai/Detection/Custom/gen_anchors.py#L37).

I have two questions:
1) what is the purpose of kmeans? the code is clear but it is not well documented/commented, and this make it hard to follow

2) does it make sence to implement kmeans from scratch, instead of using sklearn, which provide a more efficient, well documented and better maintained than this?

Would like to hear from you @OlafenwaMoses

**Comment by OlafenwaMoses** [comment]
- The **k-means** implementation is for computing the best-fit range of anchor values that generalize on all the coordinates values of the annotations in the dataset.
- This **k-means** implementation was originally implemented [here](https://github.com/experiencor/keras-yolo3/blob/e1e6073fc49839d027e58487f682de7c8a4184be/gen_anchors.py#L50)  in the [keras_yolov3](https://github.com/experiencor/keras-yolo3) project.

**Comment by rola93** [comment]
Oh good to know. I'll put an eye on it and read it again. There are still some details I don't fully understand.

My intention is to make it faster (both training process and evaluation process) for custom object detection, so I'm currently reading the code related to it. Some parts are not well documented, which make it a little hard to follow for me.

I already found some unnecesary operations, (I even write some PRs on them) which are a waste of resources (time, computation and memory).

If there is some kind of map or high level explanation on how the model is trained, it'd be great. (for instance, I don't know what exactly is an anchor, or what is the logic behind decode_network operation).

Thanks for your answer!

**Comment by plenari** [comment]
Is kmeans before or after resize the picture?

**Comment by rola93** [comment]
It's defined [here](https://github.com/OlafenwaMoses/ImageAI/blob/7a8a6012a066dc50b5db64f12885aaddd96f4e97/imageai/Detection/Custom/gen_anchors.py#L37) I'm not sure when is it exaclty used :/
